Python/optimizer_analysis.c

```diff
@@ -1,50 +1,1068 @@
+{
+    return _Py_UOpsSymType_New(ctx,NULL);
+}
+
+// Takes a borrowed reference to const_val.
+static inline _Py_UOpsSymType*
+sym_init_const(_Py_UOpsAbstractInterpContext *ctx, PyObject *const_val, int const_idx)
+{
+    assert(const_val != NULL);
+    _Py_UOpsSymType *temp = _Py_UOpsSymType_New(
+        ctx,
+        const_val
+    );
+    if (temp == NULL) {
+        return NULL;
+    }
+    sym_set_type(temp, TRUE_CONST, 0);
+    return temp;
+}
+
+static _Py_UOpsSymType*
+sym_init_push_null(_Py_UOpsAbstractInterpContext *ctx)
+{
+    if (ctx->frequent_syms.push_nulL_sym != NULL) {
+        return ctx->frequent_syms.push_nulL_sym;
+    }
+    _Py_UOpsSymType *null_sym = sym_init_unknown(ctx);
+    if (null_sym == NULL) {
+        return NULL;
+    }
+    sym_set_type(null_sym, NULL_TYPE, 0);
+    ctx->frequent_syms.push_nulL_sym = null_sym;
+    return null_sym;
+}
+
+static inline bool
+sym_is_type(_Py_UOpsSymType *sym, _Py_UOpsSymExprTypeEnum typ)
+{
+    if ((sym->types & (1 << typ)) == 0) {
+        return false;
+    }
+    return true;
+}
+
+static inline bool
+sym_matches_type(_Py_UOpsSymType *sym, _Py_UOpsSymExprTypeEnum typ, uint64_t refinement)
+{
+    if (!sym_is_type(sym, typ)) {
+        return false;
+    }
+    if (typ <= MAX_TYPE_WITH_REFINEMENT) {
+        return sym->refinement[typ] == refinement;
+    }
+    return true;
+}
+
+static uint64_t
+sym_type_get_refinement(_Py_UOpsSymType *sym, _Py_UOpsSymExprTypeEnum typ)
+{
+    assert(sym_is_type(sym, typ));
+    assert(typ <= MAX_TYPE_WITH_REFINEMENT);
+    return sym->refinement[typ];
+}
+
+
+static inline bool
+op_is_end(uint32_t opcode)
+{
+    return opcode == _EXIT_TRACE || opcode == _JUMP_TO_TOP;
+}
+
+static inline bool
+op_is_guard(uint32_t opcode)
+{
+    return _PyUop_Flags[opcode] & HAS_GUARD_FLAG;
+}
+
+static inline bool
+op_is_pure(uint32_t opcode)
+{
+    return _PyUop_Flags[opcode] & HAS_PURE_FLAG;
+}
+
+static inline bool
+op_is_bookkeeping(uint32_t opcode) {
+    return (opcode == _SET_IP ||
+            opcode == _CHECK_VALIDITY ||
+            opcode == _SAVE_RETURN_OFFSET ||
+            opcode == _RESUME_CHECK);
+}
+
+static inline bool
+op_is_specially_handled(uint32_t opcode)
+{
+    return _PyUop_Flags[opcode] & HAS_SPECIAL_OPT_FLAG;
+}
+
+static inline bool
+is_const(_Py_UOpsSymType *expr)
+{
+    return expr->const_val != NULL;
+}
+
+static inline PyObject *
+get_const(_Py_UOpsSymType *expr)
+{
+    return expr->const_val;
+}
+
+
+static int
+clear_locals_type_info(_Py_UOpsAbstractInterpContext *ctx) {
+    int locals_entries = ctx->frame->locals_len;
+    for (int i = 0; i < locals_entries; i++) {
+        _Py_UOpsSymType *new_local = sym_init_unknown(ctx);
+        if (new_local == NULL) {
+            return -1;
+        }
+        sym_copy_immutable_type_info(ctx->frame->locals[i], new_local);
+        ctx->frame->locals[i] = new_local;
+    }
+    return 0;
+}
+
+static inline int
+emit_i(uops_emitter *emitter,
```

Comment:
Note to self: I need to do a deep dive on the emit infrastructure and how it's used. I have half a feeling that we may be able to avoid having to allocate a temporary output buffer -- we may be able to prove that we can safely just overwrite the input buffer (i.e. the output provably never overtakes the input).

Reply:
We can overtake when we emit the first `_SHRINK_STACK + LOAD_CONST_INLINE` pair.